File: src/isnad/audit/canonical.py
"""Canonical JSON serialization for integrity hashing.

RFC 8785 (JSON Canonicalization Scheme, JCS) specifies a deterministic JSON
serialization: object keys sorted lexicographically, no insignificant
whitespace, and minimal string escaping (only ``"``, ``\\``, and control
characters; non-ASCII is emitted as UTF-8).

ISNAD audit records contain only strings, integers, booleans, lists, and dicts
— never floats — so the JCS number-serialization rules (the only part of the
RFC that is genuinely fiddly) are trivially satisfied.  ``json.dumps`` with
``sort_keys``, the tightest separators, and ``ensure_ascii=False`` is therefore
byte-compatible with JCS for every object ISNAD emits.

This module is stdlib-only: ``hashlib`` + ``json``.
"""

from __future__ import annotations

import hashlib
import json


class MalformedLogError(Exception):
    """A tamper-evidence log line could not be parsed into a valid entry (#108).

    Raised by the on-disk readers (``chainlog._read_chain``,
    ``merkle_log.read_batch_log``) when a non-blank line is invalid/truncated
    JSON, is missing a required key, or has the wrong shape. ``index`` is the
    0-based position among non-blank lines; ``reason`` is the human-readable
    cause. Verifiers catch this and surface a structured break, so a corrupted
    or partially-written log reports "broken" and the CLI exits 1 rather than
    crashing with a traceback — which is the whole point of a tamper-evidence
    verifier: malformed input is exactly what it must survive, not die on.

    (Consolidated here from the two audit readers after #108; originally two
    duplicate definitions.)
    """

    def __init__(self, index: int, reason: str):
        self.index = index
        self.reason = reason
        super().__init__(f"malformed entry {index}: {reason}")


def canonical_json(obj: object) -> str:
    """Serialize an object to its RFC 8785-compatible canonical form."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def sha256_hex(text: str) -> str:
    """Lowercase hex SHA-256 of UTF-8 text."""
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def canonical_hash(obj: object) -> str:
    """SHA-256 over the canonical JSON of ``obj`` (the audit record hash)."""
    return sha256_hex(canonical_json(obj))

File: js/src/canonical.ts
/**
 * Canonical JSON serialization matching the Python core's
 * `json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False)`.
 *
 * ISNAD audit records contain only strings, integers, booleans, null, arrays,
 * and objects — never floats. Floats are rejected (the "never floats"
 * invariant), which is what makes this byte-compatible with RFC 8785 for every
 * object ISNAD emits: the only genuinely fiddly part of JCS (number
 * serialization) is enforced away instead of reimplemented.
 */
import { createHash } from "node:crypto";

const HEX = "0123456789abcdef";

function escapeString(s: string): string {
  let out = '"';
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c === 0x22) out += '\\"';
    else if (c === 0x5c) out += "\\\\";
    else if (c === 0x08) out += "\\b";
    else if (c === 0x0c) out += "\\f";
    else if (c === 0x0a) out += "\\n";
    else if (c === 0x0d) out += "\\r";
    else if (c === 0x09) out += "\\t";
    else if (c < 0x20) {
      out += "\\u" + HEX[(c >> 12) & 0xf] + HEX[(c >> 8) & 0xf] + HEX[(c >> 4) & 0xf] + HEX[c & 0xf];
    } else {
      // Literal, including non-ASCII (UTF-8 emitted on hash) and U+2028/U+2029
      // (Python's ensure_ascii=False does NOT escape these; neither do we).
      out += s[i];
    }
  }
  return out + '"';
}

function compareByCodePoint(a: string, b: string): number {
  const ac = Array.from(a, (c) => c.codePointAt(0)!);
  const bc = Array.from(b, (c) => c.codePointAt(0)!);
  const n = Math.min(ac.length, bc.length);
  for (let i = 0; i < n; i++) {
    if (ac[i] !== bc[i]) return ac[i] - bc[i];
  }
  return ac.length - bc.length;
}

function serialize(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "string") return escapeString(value);
  if (typeof value === "number") {
    if (!Number.isInteger(value)) {
      throw new Error(
        "canonicalize: non-integer numbers are not allowed (ISNAD audit records never contain floats)",
      );
    }
    return String(value);
  }
  if (typeof value === "boolean") return value ? "true" : "false";
  if (Array.isArray(value)) {
    return "[" + value.map(serialize).join(",") + "]";
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj);
    keys.sort(compareByCodePoint);
    return "{" + keys.map((k) => escapeString(k) + ":" + serialize(obj[k])).join(",") + "}";
  }
  throw new Error("canonicalize: unsupported type " + typeof value);
}

/** Serialize a value to ISNAD's canonical (RFC 8785-compatible) JSON form. */
export function canonicalize(value: unknown): string {
  return serialize(value);
}

/** Lowercase hex SHA-256 of UTF-8 text. */
export function sha256Hex(text: string): string {
  return createHash("sha256").update(text, "utf8").digest("hex");
}

/** SHA-256 over the canonical JSON of a value (the audit record hash). */
export function canonicalHash(value: unknown): string {
  return sha256Hex(canonicalize(value));
}
